### PR TITLE
Setup prebuild script to ease buddybuild on boarding 

### PR DESCRIPTION
**Pull request by Palleas**
Hi there!

@BasThomas showed me the app and I really like it. Since I'm a dev evangelist at buddybuild, I asked about CI and he redirected me to #183. 

I've forked the app and [on boarded it in buddybuild](https://dashboard.buddybuild.com/public/apps/59c652e2f96f890001f0aa44/build/59c660c184d8e8000188ce27#logs
), if you feel like taking buddybuild for a spin, it should work right away, the only missing stuff was adding the Apollo Codegen thingie. 

I also enabled the execution of tests, but they seem to be failing for now: https://dashboard.buddybuild.com/public/apps/59c652e2f96f890001f0aa44/build/59c668f7fc18d80001f033df

Let me know if there is anything else I can do to help!

### CODE REVIEW

**Comment by rnystrom** [comment]
This is crazy. Can’t wait to try this out. Thanks @Palleas for giving this a go!!!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
🙌 amazing @Palleas!

**Comment by Palleas** [comment]
Since we've added a `package.json`, we no longer need a prebuild to install apollo-codegen, buddybuild automatically runs `npm install`. 

Do you want to close this issue or is there anything else you'd like to have on buddybuild that I could setup as part of this PR?

**Comment by Sherlouk** [comment]
Apologies I'm not too familiar with the intricacies around BuddyBuild

So if we wanted to run the tests on multiple different simulators, where do we determine that?

We've also recently added a feature which means we need to send different builds to TestFlight and Release/App Store. Does BB have something that would help with this?

**Comment by Palleas** [comment]
No worries! Happy to answer any questions you might have, and it’s usually a good way to improve our product.

**Comment by Palleas** [comment]
No worries! Happy to answer any questions you might have, and it’s usually a good way to improve our product.

> So if we wanted to run the tests on multiple different simulators, where do we determine that?

There is a test settings panel which let you select multiple devices:

01-test-settings

> We've also recently added a feature which means we need to send different builds to TestFlight and Release/App Store. Does BB have something that would help with this?

You can automatically deploy to iTunesConnect, it won't do the actual submit part but some customers have achieved this extra step using fastlane: https://www.buddybuild.com/blog/automating-iOS-releases

> Additionally I see plans to use fastlane a little bit more for ease of doing certain actions, how seamless is the integration between the two tools? Could we ask BB to run certain lanes given certain criteria? (Linking to the previous question)

There are a few way to customize the build process: https://www.buddybuild.com/blog/customize-mobile-build-steps

Basically, you could run certain lanes based on the branch that we're buddybuild, a specific env variable, etc.. 

If you have more concrete examples, I'll be happy to add them here for you so you can see what it will look like.

**Comment by Sherlouk** [comment]
So I think the workflow is still being molded into it's final form but basically in my mind (Ryan can confirm or deny)

Around every other day, if there are changes, a ["TestFlight" build](https://github.com/rnystrom/GitHawk/blob/master/fastlane/Fastfile#L18) (not release build!) should be created, tested, and sent to our testers via TestFlight.

At any point when Ryan decides "why the hell not", a [Release build](https://github.com/rnystrom/GitHawk/blob/master/fastlane/Fastfile#L24) should be created, tested, and delivered to the App Store ready for Apple's approval. Obviously if we notice something has gone wrong, we still need a backout plan where that version is removed, and this cycle starts again creating a new build (from recent fixes), tested, delivered.

In my head, master should always be deployable to our testers to ensure by the time it's released, it's been used! The key thing here is that the same build **can not** go to TestFlight and App Store.

Great to see how easy it is to be testing on multiple simulators, and I do look forward to having it automatically tested on beta's when they come out!

**Comment by rnystrom** [comment]
Awesome! So I took the time to get this setup w/ a pretty basic scheme. Going to give it a whirl for a few weeks and see how it pans out. I have it configured to build the TestFlight scheme, would love to figure out how to take a succesful build and send it to iTunes Connect!

**Comment by rnystrom** [comment]
Also curious to see how this works w/ PRs (reporting success/failure for builds).

**Comment by Palleas** [comment]
I’m back at work next week, happy to jump on a hangout or whatever chat solution works for you if you have any question! :) 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
@Palleas could use your help really quick. I noticed that your build executes the "Generate Apollo" script build phase just fine, but I'm getting an error after changing to the Freetime - Testflight scheme:

https://dashboard.buddybuild.com/apps/59d904ff8588d60001729c4e/build/59d9136c1ef40f0001041b0b#logs

However it did succeed when using the Freetime - Release scheme

https://dashboard.buddybuild.com/apps/59d904ff8588d60001729c4e/build/59d906ac6b97d800018bad53#logs

Is there some caching issue or something?

**Comment by Palleas** [comment]
Can you try to do “rebuild without cache” for the failed build detail page? I’ll have my laptop in a few hours so I should be able to provide more help!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Sherlouk** [comment]
Just kicked off a build without cache, also may be worth sharing the scheme? I, at least, don't do anything with the scheme but it probably isn't a bad idea to have one shared!

**Comment by Sherlouk** [comment]
Build failed a few things I'm noticing:

`find: -: unknown primary or operator`

I'm hazarding a guess and thinking something doesn't like the fact that BB is adding spaces to the directory names?

`Couldn't find Apollo.framework in FRAMEWORK_SEARCH_PATHS; make sure to add the framework to your project`

This is the error that we print out, if it can't find `Apollo.framework`. I'm not familiar with buddybuild but the logs to me are saying that it extracts each of the scripts and runs them in isolation? Could be wrong but I wonder if the script we're using to locate the framework isn't being found.

--

I'd also like to add the other thought to the mix, should we not be using the "IS_CI" environment variable as defined by Buddybuild (or we can define our own) and **not** run that script in this situation? I understand it looks like I'm trying to avoid the issue, but actually -- should we be re-generating the file that drives most of our networking? I feel like that's going to cause more issues, or reversely fix an issue that makes it more difficult for us to find/someone forgot to commit a change? I'd rather not do this at all and let the build fail if something's wrong with `API.swift`?

--

As a final thought, @martijnwalraven - the issue we're having here seems to be directly related to Apollo's run script. Have you or any of your community had issues or success running on BB? (Trying to see if it's something in their architecture?)

**Comment by martijnwalraven** [comment]
@Sherlouk: I've never used buddybuild, but the Apollo iOS tests run without any issue on Travis CI.

**Comment by rnystrom** [comment]
@Palleas another question for you while sorting this out:

How can we use buddybuild along w/ TestFlight? Some ideas that would be nice:

- Download build and manually upload to iTunes Connect
- Automatically upload to iTunes Connect on a schedule (e.g. every 3 days when build is green)

**Comment by rnystrom** [comment]
Hmph, Release config also failed :/

https://dashboard.buddybuild.com/apps/59d904ff8588d60001729c4e/build/59db81cc85619400018f0390#logs

**Comment by Sherlouk** [comment]
@rnystrom I'm still thinking we just skip that step if it's being run from CI (Using an env variable) see my comment above, but there's no reason we should be doing it.

**Comment by rnystrom** [comment]
Ya that's true, totally not necessary during CI

**Comment by Sherlouk** [comment]
Also @rnystrom looks like you can set up an iTC release schedule:

https://docs.buddybuild.com/deployments/ios/itunes_connect.html

Looks like it's limited to days of the week rather than every x days, maybe a new feature request 😅 

For the time being though we could just do releases every Tuesday and Friday? (or any other day combination) Obviously latest successful master build

**Comment by BasThomas** [comment]
Yes please! 🙏

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Palleas** [comment]
It definitely looks like something's up with the `find` command... I'm able to reproduce a similar issue with `find . -name  -maxdepth 1` (note the empty name argument). 

Looking at the find command, I wonder if the cause could be the use of `"`?

screen shot 2017-10-08 at 8 39 55 pm

Can we try and replace with the following?

```
APOLLO_FRAMEWORK_PATH="$(eval find $FRAMEWORK_SEARCH_PATHS -name \"Apollo.framework\" -maxdepth 1)"
```

That doesn't explain why it would only fail sometimes though 🤷‍♂️ 

Thank you @Sherlouk for pasting the link to our documentation, we indeed don't support scheduling "Every x day", but that's an interesting idea!

**Comment by weyert** [comment]
Yes, I raised that request via mail about be able to do builds every x days or weeks. As I only wanted to make a new build each TestFlight build expires. Only this wasn't supported yet. I hope they will consider supporting it.

Said that, I really enjoyed using BuddyBuild when I was playing around with it 👍

**Comment by Palleas** [comment]
> Said that, I really enjoyed using BuddyBuild when I was playing around with it 👍

Really happy to read that!

**Comment by Sherlouk** [comment]
Going to close this now that we have BuddyBuild functional, thanks for your help @Palleas!
